### ISSUE TITLE
[PROPOSAL] standard dataset facet for DBMS objects

### ISSUE THREAD

**Issue by davidjgoss** [OpenLineage/OpenLineage#2774]
**Purpose:**
Many datasets in an OpenLineage landscape are likely to be objects (tables, views) within a DBMS (in which I'm including Postgres, MySQL, Oracle etc, plus cloud warehouses like Snowflake etc).

Such objects have properties which may be of interest to an OpenLineage-ingesting system, but would not be applicable to other kinds of dataset (blob storage, APIs, etc). Specifically:

- Hierarchy - OpenLineage models a dataset name as a single string which is reasonable. DBMS objects tend to have a name derived from a hierarchy e.g. `database` -> `schema` -> `table` become `database.schema.table`. A consuming system may want to model those parts separately to e.g. make the hierarchy navigable in a tree fashion by a user, or assist searching. They could parse the single name string, but this would be fussy and error-prone - just splitting on the period is not enough for some object names, where parts may have to have escape characters added to ensure they are unambiguous. Also, although `database.schema.table` is the most common, several vendors have a different number/naming of items e.g. MySQL, Databricks.
- Object type - Both tables and views can be represented as datasets; it may benefit a consuming system's user to know which type a dataset is.

**Proposed implementation**
I'd propose a new `dbmsObject` facet, applicable to both input and output datasets, and static datasets.

Example dataset:
```json
{
  "namespace": "snowflake://my-account.eu-central-1",
  "name": "MY_DATABASE.MY_SCHEMA.\"InterestinglyNamedTable\"",
  "facets": {
    "dbmsObject": {
      "_producer": "...",
      "_schemaURL": "...",
      "objectType": "TABLE",
      "hierarchy": [
        {
          "name": "MY_DATABASE"
        },
        {
          "name": "MY_SCHEMA"
        },
        {
          "name": "InterestinglyNamedTable"
        }
      ]
    }
  }
}
```

Schema:
```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "...",
  "type": "object",
  "properties": {
    "type": {
      "type": "string",
      "description": "The type of object in the DBMS. Valid values may include `TABLE` and `VIEW`."
    },
    "hierarchy": {
      "type": "array",
      "description": "The names of items in the hierarchy within the namespace (e.g. database, schema, table), from highest to lowest level",
      "items": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string",
            "description": "The name of the item in the hierarchy"
          }
        }
      }
    }
  }
}
```

**Comment by mobuchowski** [comment]
>Hierarchy - OpenLineage models a dataset name as a single string which is reasonable. DBMS objects tend to have a name derived from a hierarchy e.g. database -> schema -> table become database.schema.table. A consuming system may want to model those parts separately to e.g. make the hierarchy navigable in a tree fashion by a user, or assist searching. They could parse the single name string, but this would be fussy and error-prone - just splitting on the period is not enough for some object names, where parts may have to have escape characters added to ensure they are unambiguous.

I agree, would add that the hierarchy ranges from just two elements in MySQL and probably some other DBs, to four in Snowflake. 

>Object type - Both tables and views can be represented as datasets; it may benefit a consuming system's user to know which type a dataset is.

Yes - I think that's very valid.

```
      "enum": [
        "TABLE",
        "VIEW"
      ]
```

I've learned that putting enums in schema is not a great idea since it prevents us from being forward-compatible. We might want to add some future value there (materialized view perhaps?) and if the consumers are locked into having only those values, they sometimes can't even deserialize it.

**Comment by jakub-moravec** [comment]
@davidjgoss as you mention ..

> although database.schema.table is the most common, several vendors have a different number/naming of items e.g. MySQL, Databricks.

I'll add one more - Teradata, which doesn't have schemas. Some catalogs choose to "skip" the level to reflect the structure of the database, but others prefer consistency and will, for example, use a database name also as a schema name.

It would be more consumer-friendly to label the hierarchy segments by types so that each tool can represent the hierarchy and construct qualified names as they see fit. 

E.g. 

```
{
  "namespace": "snowflake://my-account.eu-central-1",
  "name": "MY_DATABASE.MY_SCHEMA.\"InterestinglyNamedTable\"",
  "facets": {
    "dbmsObject": {
      "_producer": "...",
      "_schemaURL": "...",
      "objectType": "TABLE",
      "hierarchy": [
        {
          "objectType": "DATABASE",
          "name": "MY_DATABASE"
        },
        {
          "objectType": "SCHEMA",
          "name": "MY_SCHEMA"
        },
        {
          "objectType": "TABLE"
          "name": "InterestinglyNamedTable"
        }
      ]
    }
  }
}
```

**Comment by dolfinus** [comment]
PR with `HierarchyDatasetFacet`: #3747
